### ISSUE TITLE
Questions for pytorch_spl_loss.py

### ISSUE THREAD

**Issue by victorca25** [ssarfraz/SPL#2]
Hello!

While checking the Pytorch version, I noticed that SPLoss() does not use the "weight" as the SPL_ComputeWithTrace() version does. I just want to be sure, can these variables be removed?

Do SPLoss() and SPL_ComputeWithTrace() give the same results? I'm not sure I understand how is the trace of the transpose happening in SPLoss().

Contrary to most loss functions, the GPL and CPL losses are coming up with negative values. Is that expected? The model is not becomming unstable or anything, I just didn't expect that.

Lastly, I may have some error somewhere (maybe related to the losses signs), but my result images are coming over exposed (too bright) and a bit blurry. I'm using a pretrained model for SR (was trained using a GAN before) as a basis and I'm seeing it degrade almost inmediatelly. My images are in the [0,1] range, so I'm not using the conversions, in case that makes a difference.

Thanks for the very interesting work, I've been looking for alternative loss functions and this is very promising!

**Comment by ConstantinSeibold** [comment]
Hello victorca25.

1. The variable 'weight' was originally intended to weigh color channels differently. In our experiments, we found that an equal weight between all channels gives the best results. As such, this variable is a leftover from that time and can be removed.

2. Both loss versions equate to the cosine similarity of rows/columns. While in 'SPL_ComputeWithTrace()' this is achieved using the trace (sum over the diagonal) of matrix multiplication of L2-normalized input/target rows/columns, 'SPLoss()'  L2-normalizes the rows/columns, performs piece-wise multiplication of the two tensors and then sums along the corresponding axes. The latter variant, however, needs less operations since it can be performed batchwise and, thus, is the preferred variant.

3. Indeed, the loss function are expected to be negative. As we want to maximize the similarities, we, as stated in the paper, minimize the negative of the similarities.

4. We didn't try finetuning a GAN-pretrained model using the SPL. It might be the case that since the SPL pushes the network towards a different objective than a prior discriminator did, the model might not learn effectively.
Do you face these issues as well when not using a pretrained model? When we trained the SR-models the initial epochs also looked slightly blurry and got better over time. 

Also, since we used tanh as generator output, we rescaled the tensors to a [0,1] range from [-1,1] and also used [-1,1] ranged input images to be able to use the rgb-yuv conversion in the CP component. What activation function are you using for your output layer? Not using any or using ReLU might lead to bright images as nothing caps your outputs inside the [0,1]-range and your values might overflow when you transfer it back to opencv/Pillow for visualization. 

As a side note, we used the tensorflow version for our SR experiments. While there shouldn't be any differences between the TF and Pytorch loss functions, the change in frameworks might have an effect.

**Comment by victorca25** [comment]
Thank you very much for the very detailed answer and guidance!
For 1. I imagined the weight in both cases was not affecting much, since they were set to 1 in all three cases, but it's very interesting to know about those additional experiments. 
For 2. I kind of get it now, but I have to keep testing it to understand better. Interestingly, by swapping SPLoss() with SPL_ComputeWithTrace() in both GPLoss() and CPLoss(), the validation images no longer look much more brighter than they should be (still blurry, but it did change the results in that aspect). 
3. Excellent!
4. That does make sense, even to explain the blurriness. It was taking long when training from scracth to see some results, so I though of using this model as pretrained to jumpstart the model instead of starting from random noise, but I'll keep testing and give it more time. 

Currently I have no activation in the generator output, I will add a tanh activation and continue testing. I'm using OpenCV to work with the images which I normalize to the [0,1] range. To be sure, in order for to_YUV() to work correctly, images should be in the [0,1] range or [-1,1]?

I did try to use the Tensorflow code as reference, but the structure is different enough to try to compare 1:1 to a PyTorch version. I'll probably try to do a side by side comparison of the execution of both cases to see if anything differs. 

I'll write again if I find anything.

Cheers and again, thanks!

**Comment by ssarfraz** [comment]
Hi
Would be best to scale your inputs and targets in -1 1  and use tanh on the output layer. Then use spl with deafault as it will convert all three to 0-1 before - yuv

**Comment by victorca25** [comment]
I've added the tanh activation to the model output and normalized all images on the [-1, 1] range and started a new training session from scratch to see what happens in this case. So far it's still black outputs with a tiny bit of blurry colors and shapes, but I'll let it run longer.

Another question just out of curiosity, is there any particular reason for using the BT.601 YUV constants when converting from RGB, say, in comparison to using BT.709 (HDTV)? I imagine results should be similar, since the differences are only constants, but I was wondering if there was a particular reason for that choise. 

Thanks!

**Comment by ssarfraz** [comment]
no reason we just used the defaults provided in tensorflow.  by the way it shouldn't be black outputs..it should pick the structure pretty fast. which generator and data you are working with?

**Comment by victorca25** [comment]
I'm using the residual in residual dense block network (RRDBNet) from ESRGAN, which is pretty similar to the SRGAN network (the only modification is the tanh activation at the end) and part of the DIV2K dataset for training. 

And you are correct, I restarted training once again and forgot to change from SPL_ComputeWithTrace() back to SPLoss() and I got colors pretty fast on the next image validation after the switch (1k iterations): 
![image](https://user-images.githubusercontent.com/41912303/64921279-96318b00-d7c1-11e9-9529-2c3818e42749.png)

This is the target image 
![image](https://user-images.githubusercontent.com/41912303/64921292-d264eb80-d7c1-11e9-8d66-10fcf1842871.png).

The magnitude of the losses are pretty similar in both cases (some representative values are: loss_gpl: -3.5991e+01 and loss_cpl: -1.2673e+02), I'm not sure what's making the difference in using one versus the other.

**Comment by ssarfraz** [comment]
hi, your output still seems a bit weird to  me, May be check it with the official tensorflow train scripts provided.  I am gonna close this issue now, hope it helped!

**Comment by victorca25** [comment]
Hello! I wanted to update, but didn’t want to create a new issue for it. Now the brightness is fine, I got it working for both cases, [0,1] range and [-1,1] range, the help was very helpful. For both cases results are the same after training for a long time, the resulting models look almost exactly like a model trained with regular L1 loss (blurry) it is not getting additional details from there.

Any additional suggestions or tip where to look at? Cheers and thanks again!
